File: data_structures/graph/directed/cyclic/directed_cyclic_graph.py
class Graph:
    def __init__(self, undirected=False):
        self.isUndirected = undirected
        self.nodes = []
        self.length = 0

    def get_length(self):
        """
        Returns the current length (e.g. number of Nodes) of the Graph
        """
        return self.length

    def set_length(self, value):
        """
        Sets a new length (eg.g number of Nodes) of the Graph
        """
        self.length = value

    length = property(get_length, set_length)

    def add_node(self, value):
        """
        Adds a new value to the Graph inside a Node class
        """
        self.nodes.append(Node(value))
        self.length = self.length + 1

    def remove_node(self, value):
        """
        Removes a value and its Node from the Graph and removes any corresponding edges that
        connect it to another Node in the Graph
        """
        self.nodes = [node for node in self.nodes if node.value != value]

        for node in self.nodes:
            node.edges = [node for node in node.edges if node.value != value]

        self.length = self.length - 1

    def get_nodes(self, value):
        """
        Returns an array with any and all Nodes found that satisfy the value parameter
        """
        nodes_list = []

        for node in self.nodes:
            if node.value == value:
                nodes_list.append(node)

        return nodes_list

    def add_edge(self, value1, value2):
        """
        Adds an edge from the Node containing value1 to the Node containing value 2
        """
        node1 = None
        node2 = None

        for node in self.nodes:
            if node.value == value1:
                node1 = node
            elif node.value == value2:
                node2 = value1

        node1.edges.append(node2)

        if self.isUndirected:
            node2.edges.append(node1)


class Node:
    def __init__(self, value):
        self.value = value
        self.edges = []

File: data_structures/graph/directed/cyclic/directedCyclicGraph.js
class Graph {
  constructor(undirected = false) {
    this.isUndirected = undirected;
    this.nodes = [];
    this.length = 0;
  }

  get length() {
    /**
     * Returns the current length (e.g. number of Nodes) of the Graph
     */
    return this.length;
  }

  addNodes = (value) => {
    /**
     * Adds a new value to the Graph inside a Node class
     */
    this.nodes.push(new Node(value));
    this.length++;
  };

  removeNode = (value) => {
    /**
     * Removes a value and its Node from the Graph and removes any corresponding edges that
     * connect it to another Node in the Graph
     */
    this.nodes.filter((node) => node.value !== value);
    this.nodes.forEach((node) => {
      node.edges.filter((node) => node.value !== value);
    });
    this.length--;
  };

  getNode = (value) => {
    /**
     * Returns an array with any and all Nodes found that satisfy the value parameter
     */
    return this.nodes.filter((node) => node.value === value);
  };

  addEdge = (value1, value2) => {
    /**
     * Adds an edge from the Node containing value1 to the Node containing value 2
     */
    const node1 = this.nodes.find(value1);
    const node2 = this.nodes.find(value2);

    node1.edges.push(node2);

    if (this.isUndirected) {
      node2.edges.push(node1);
    }
  };
}

class Node {
  constructor(value) {
    this.value = value;
    this.edges = [];
  }
}

module.exports = {
  Graph,
  Node,
};
